ruby: Enable built-in Ruby Mode

lunaryorn opened the pull request:
Add built-in Ruby Mode.  Keybindings are added for both modes, and there are two extra bindings for the built-in mode which have no equivalent in Enh Ruby Mode to my knowledge.

This change enables Ruby Mode by default on Emacs 24.4 and newer where it was significantly improved, and otherwise keeps using Enh Ruby Mode.  That's a breaking change, obviously, and I'm fine with keeping enh ruby mode as default, too, so if you'd like to have that changed, just tell me :relaxed:  I think, though, that as of 24.4 the built-in Ruby Mode is much better than Enh Ruby Mode, but YMMV obviously.

In any case, there's a variable to explicitly enable/disable the built-in mode, of course.

That's my first major change to a layer, and I'm not sure whether it's all correct.  Please review carefully :blush:

TheBB commented:
Any heavy Rubyistas want to chime in on this?

robbyoconnor commented:
I would like a configuration option rather than forcing this.

syl20bnr commented:
@bru are you still using spacemacs ? I know we were divided about replacing the built-in ruby mode but cannot remember for what reason exactly, any objection to re-enable the ruby built-in package ?

robbyoconnor commented:
I use Ruby -- which is why I chimed in :)

dsdshcym commented:
There seems to be some problems with company mode in this commit. At least, I couldn't use company mode in ruby mode when I was trying these changes.

For other parts, I think it's great to have an option for these. I personally don't like `enh-ruby-mode`. Since it has poor indentation supports for ruby.

lunaryorn commented:
@robbyoconnor Read the diff and my comment again :wink: 

@dsdshcym Thank you very much for testing :relaxed:  I’ve indeed made a mistake in the Company setup.  It's fixed now, would you mind to try again? :blush:

robbyoconnor commented:
> @robbyoconnor Read the diff and my comment again :wink:
DOH!

bru commented:
hi all, @syl20bnr yes I still use spacemacs, although I didn't get much time to contribute these past months :-/
IIRC the reason for switching to enh-ruby-mode was better syntax support (for ruby 2), but that was a long time ago.
Honestly if the built-in ruby mode will help speeding up things in ruby then it gets my vote, I still find editing ruby files frustratingly slower than on vim.

lunaryorn commented:
@bru @syl20bnr I should like to emphasise that this pull request is not about _replacing_ Enh Ruby Mode.  I've chosen to make the builtin mode the default, because I think it's superior, but that's just my personal opinion, and YMMV.  I don't cling to that choice.  As far as I'm concerned we can also keep Enh Ruby Mode the default, and with that choice this pull request doesn't change anything at all for users of this layer.

Apparently my choice of a different default has distracted from the real purpose of this PR.  I've reverted it to put this PR back to its real purpose:  **Add equal support** for the built-in Ruby Mode, which is currently lacking entirely.  Now this PR just adds minor modes and key bindings for the built-in mode, but doesn't actually enable it.  For users for this layers _nothing_ will change with this PR, _unless_ they explicitly change the new `ruby-use-built-in-ruby-mode` option to a non-nil value.

I'd really love to see this base-line merged, because the experience with Enh Ruby Mode is very much inferior to the built-in Ruby Mode in Emacs 25.

lunaryorn commented:
Inviting @dgutov and @bbatsov, for their feedback.

bbatsov commented:
As a "heavy" Rubyist I think it's pretty odd idea to replace one of the best built-in Emacs modes, with a mode that's not maintained that actively, relies on an external tool (and is known for some shortcomings). 

No idea, who made this call, but it's time for sanity to prevail. Several friends of mine complained about this default when they were switching from vim to spacemacs and they were all quite happy with `ruby-mode`.

igas commented:
:+1: I use forked version of layer to get rid of `enh`, but it's hard to keep it up to date with upstream layer.
@lunaryorn it also need change to rails layer, it has hook for enh.

dgutov commented:
I use the built-in mode, and I intend to continue fixing it (although admittedly I haven't done much of that for the last months). I've read good things about `enh-ruby-mode`, though.

robbyoconnor commented:
> :+1: I use forked version of layer to get rid of enh, but it's hard to keep it up to date with upstream layer.
> @lunaryorn it also need change to rails layer, it has hook for enh.

Yeah this needs to be done.

dsdshcym commented:
@lunaryorn There are still some problems with company mode. I can auto-complete some builtin functions. But I can't auto-complete variables and other functions. 

Maybe there are still some works to do with it. But I'm very happy to use the built-in ruby mode.

bru commented:
@bbatsov to summarise, I'm guilty for making the initial call for `enh` in the early days of spacemacs; that was based solely on my personal (very limited) experience with emacs, and in the hope that the enhanced ruby mode would have been maintained, kept relevant and improved over the builtin.
As mentioned already, I was an emacs newbie then and I mostly still am now, so by all means :+1: from me to switching the default to the built-in mode if that's going to improve experience and make it easier for users to adopt and enjoy spacemacs.

lunaryorn commented:
Jesus, folks!  :unamused: This isn't about _switching_, at least not yet.  I'd just like to have equal support for Ruby Mode, and I'd appreciate if you'd test the Ruby support and provide feedback about the implementation instead of discussing which one is better.  

That's ultimately a judgement call anyway, and I'm not in a position to decide.  And once this one is merged, it's easy enough to change the default.

Please, let's keep this discussion out of here.

@robbyoconnor @igas I've updated the rails layer, but I don't use rails.  I'd appreciate if the burden of testing rails wasn't put upon my shoulders.

@dsdshcym I've fixed another problem. :blush:  Sorry to bug you, but would you mind to test yet again?

bru commented:
sorry @lunaryorn, just providing a bit of background :) I'll also try to run it on a rails project asap

TheBB commented:
But if the built-in mode is better, we can make this about switching. At the very least, about switching the default. It's also a matter of dev burden to keep two modes working with the same feature sets.

lunaryorn commented:
@bru I'm sorry, I didn't mean you specifically.  I'm just very disappointed to see everyone discussing whether one or the other is better, but zero to no feedback on the actual changes.

@TheBB  I'd not like to see this PR being delayed by a lengthy discussion about the best default, and I'm not going to take responsibility—and thus blame—for changes to the default.

----

See, I just don't care for the default. It's just one option away with this change.  But I do care about good Ruby Mode support, and I'd really love to see this PR being merged fast.  

After all, at some point I need to get my own work done, you know :)

trishume commented:
:+1: on this PR
and also :+1: on making it the default (in a separate PR so this one can get merged soon)

One of the reasons I switched away from Spacemacs was how laggy enh-ruby-mode is, especially when working over an NFS connection into a Vagrant box.

bru commented on layers/+lang/ruby/README.org:
```diff
@@ -24,6 +24,16 @@ To use this contribution add it to your =~/.spacemacs=
+
+#+BEGIN_SRC emacs-lisp
+  (defun dotspacemacs/user-init ()
+    (setq-default ruby-use-built-in-ruby-mode nil)
```
I think you mean `t` instead of `nil` here ;)

lunaryorn replied:
Fixed :relaxed:

bru commented:
@lunaryorn finally had a chance to look through the code. lgtm, apart from an issue in the README (see comment above). :+1: and let's discuss later what to do about the default.

dsdshcym commented:
@lunaryorn Company-mode works fine now BTW. Thanks for your work! I'm really looking forward to see this PR being merged. :smile:

robbyoconnor commented:
@trishume -- you should come back -- we have cookies now!

syl20bnr commented:
Thank you :heart: 
I'm not comfortable with a variable to enable the _built-in_ mode so I reversed it. The built-in ruby mode is enabled by default and the variable `ruby-enable-enh-ruby-mode` must be set to `t` to replace the built-in mode, often we use the verb `enable` for layer variable but I'm not sure it is appropriate, I'll check the usage of `use` verb and revert it to `use` if needed. Related commit: https://github.com/syl20bnr/spacemacs/commit/6e59240
Cherry-picked into `develop` branch, you can safely delete your branch.

lunaryorn commented:
@syl20bnr Thank you very much! :heart_eyes:

bbatsov commented:
:+1: Great!

kingcons commented:
:confetti_ball:
